File: scripts/check-symbols.py
#!/usr/bin/env python3
"""Reliable narrow check for the recurring bug class: a SCREAMING_CASE lookup
table used as NAME[...] but never declared. No cleaning, no parsing — a raw
substring test for the declaration, which cannot produce false negatives."""
import re, sys
src = open(__import__('os').path.join(__import__('os').path.dirname(__file__),'..','apps','web','src','App.jsx')).read()
used = set(re.findall(r'\b([A-Z][A-Z0-9_]{2,})\s*\[', src))
missing = sorted(n for n in used
                 if not re.search(r'\b(?:const|let|var)\s+' + n + r'\b', src)
                 and not re.search(r'\b' + n + r'\s*[:,]', src))
if missing:
    print("UNDEFINED LOOKUP TABLES:", ", ".join(missing)); sys.exit(1)
print(f"symbol check: clean ({len(used)} lookup tables verified)")

# The artifact runner evaluates App.jsx as a plain script, where import.meta is
# a parse error. Keep it out permanently.
import re as _re
_src = open("apps/web/src/App.jsx", encoding="utf-8").read()
_code = _re.sub(r"/\*[\s\S]*?\*/", "", _src)
_code = _re.sub(r"(?m)//.*$", "", _code)
if "import.meta" in _code:
    print("FAIL: import.meta is not allowed in App.jsx (breaks script-mode artifact rendering)")
    raise SystemExit(1)
print("script-mode check: no import.meta")

# The Tailwind CDN script warns in the console, ships the whole engine to every
# visitor and cannot be cached or versioned. The app compiles its own CSS.
_html = open("apps/web/index.html", encoding="utf-8").read()
if "cdn.tailwindcss.com" in _html:
    print("FAIL: apps/web/index.html loads Tailwind from the CDN; build it instead")
    raise SystemExit(1)
if "index.css" not in open("apps/web/src/main.jsx", encoding="utf-8").read():
    print("FAIL: main.jsx does not import the compiled stylesheet")
    raise SystemExit(1)
print("stylesheet check: Tailwind is compiled, not CDN")

File: scripts/check-symbols.mjs
import fs from "node:fs";
import path from "node:path";
import { fileURLToPath } from "node:url";

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const appJsxPath = path.join(__dirname, "..", "client", "src", "App.jsx");
const src = fs.readFileSync(appJsxPath, "utf-8");

const matches = [...src.matchAll(/\b([A-Z][A-Z0-9_]{2,})\s*\[/g)].map((m) => m[1]);
const used = new Set(matches);
const missing = [...used].filter((n) => {
  const declRegex = new RegExp(`\\b(?:const|let|var)\\s+${n}\\b`);
  const propRegex = new RegExp(`\\b${n}\\s*[:,]`);
  return !declRegex.test(src) && !propRegex.test(src);
}).sort();

if (missing.length > 0) {
  console.error("UNDEFINED LOOKUP TABLES:", missing.join(", "));
  process.exit(1);
}
console.log(`symbol check: clean (${used.size} lookup tables verified)`);

// Check import.meta in App.jsx
const stripped = src
  .replace(/\/\*[\s\S]*?\*\//g, "")
  .replace(/\/\/.*$/gm, "");

if (stripped.includes("import.meta")) {
  console.error("FAIL: import.meta is not allowed in App.jsx (breaks script-mode artifact rendering)");
  process.exit(1);
}
console.log("script-mode check: no import.meta");

// Tailwind check
const html = fs.readFileSync(path.join(__dirname, "..", "client", "index.html"), "utf-8");
if (html.includes("cdn.tailwindcss.com")) {
  console.error("FAIL: client/index.html loads Tailwind from the CDN; build it instead");
  process.exit(1);
}
const mainJsx = fs.readFileSync(path.join(__dirname, "..", "client", "src", "main.jsx"), "utf-8");
if (!mainJsx.includes("index.css")) {
  console.error("FAIL: main.jsx does not import the compiled stylesheet");
  process.exit(1);
}
console.log("stylesheet check: Tailwind is compiled, not CDN");
